### ISSUE TITLE
Booting Halium on ubports-boot

### ISSUE THREAD

**Issue by UniversalSuperBox** [Halium/projectmanagement#40]
With just a few tweaks to both the rootfs and ubports-boot, it's pretty easy to run Halium on ubports-boot. 

## Building ubports-boot

You'll need to pull ubports-boot and get it to build in order to test the installation of Halium on it.

First, clone https://github.com/ubports/ubports-boot.git to `halium/ubports-boot` in your tree, right alongside `hybris-boot`.

Then, change into `ubports-boot` and pull a few unmerged changes:

```
git pull https://github.com/UniversalSuperBox/ubports-boot.git stop-halium-clash --rebase
git pull https://github.com/UniversalSuperBox/ubports-boot.git halium-layout --rebase
git pull https://github.com/UniversalSuperBox/ubports-boot.git not-always-ro --rebase
```

Then find the line `# Specific projects for Halium` in `build/core/main.mk`. Make the next lines look like this:

```
subdirs += \
        halium/hybris-boot \
        halium/ubports-boot
```

Now, unfortunately, you'll need to edit your device's `fstab`. Change the source for `/data` to `/dev/mmcblk[x]p[y]`. This is something we can easily solve since we have fixup-mountpoints.

Now, build mkbootimg and ubports-boot. 

## Installation.

First, flash ubports-boot to the device. It's time for the rootfs!

Install the system.img and rootfs.img to the device with your preferred method. 

While you're in recovery, do `touch /data/writable_image` so that the rootfs stays read-write when it gets mounted.

Once booted, check if `/android`, `/userdata`, and `/var/lib/lxc/android/rootfs` exist. If they don't, create them. Now we'll remove some customization and let ubports-boot do it for us.

Let's disable some units first:

```
systemctl disable system.mount
systemctl disable android-mount
rm /etc/systemd/system/systemd-udevd.service.d/10_copy-device-udev-rules.conf
```

Then, replace `/var/lib/lxc/android/pre-start.sh` with [this](https://gist.github.com/UniversalSuperBox/9421e4f1350e8959d2205cb3ad14a944). Make sure it's executable!

Now, we'll change a setting on `lxc@android.service`. Run `systemctl edit lxc\@android.service`. Paste the following into the editor:

```
[Install]
WantedBy=multi-user.target systemd-udevd.service
```

Now hopefully udev will wait until after the container starts. Reboot safely with `/sbin/reboot` and watch the magic happen!

**Comment by LNJ2** [comment]
Maybe this should be added to the documentation?

**Comment by UniversalSuperBox** [comment]
No, we're not yet sure exactly how to proceed in the initrd space.

**Comment by ZeroPointEnergy** [comment]
I tested this successfully for the Samsung Galaxy S7. Apart from a few minor points everything seams to work flawless:
- I needed to add console=tty0 to the kernel command line because my board sets console=ram and initrfamfs complained, something the hybris init had no problem with.
- In order for the basic tests to run I had to create a symlink  /system -> /android/system/

**Comment by LNJ2** [comment]
On the yuga I didn't need the first step, but I also had the problem with the symlink.

**Comment by UniversalSuperBox** [comment]
Just added the instructions based on this feedback, thanks

**Comment by walidham** [comment]
in my case i must run this command to link vendor folder
ln -s /android/system/vendor /vendor
